rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -1750,10 +1702,9 @@ extern "C"
             {
                 void *data = guard_conditions->guard_conditions[i];
                 GuardCondition *guard_condition = (GuardCondition*)data;
-                if(!guard_condition->getHasTriggered())
-                {
-                    guard_conditions->guard_conditions[i] = 0;
-                }
+                // Guard conditions persist (don't set to null)
+                // We call getHasTriggered to reset the trigger state.
```

Comment:
@wjwwood wrote a rather long post about guard conditions: https://github.com/ros2/rcl/pull/33#issuecomment-198531721

I don't think that it's necessary to signal which guard condition triggered to the caller of wait. We use this pattern for subscriptions, services, and clients that get triggered so that the executor knows which callback to call, but guard conditions have no corresponding callback (they simply wake up the blocking call to wait).